ls1intum/Artemis
# [Feature] - course statistics

### Checklist

### Motivation and Context


We want to have a statistics page for a specific course

### Description

added a new course statistics page where one can see metrics about the specific course. Can be accessed by clicking on the _Statistics_ Button in the course detail view
For the average scores, only ten exercises will be shown in once. The user can click on the right or left arrow to adapt the exercises shown
Also, major server side statistics logic got refactored and moved from the StatisticsService into the StatisticsRepository


### Steps for Testing


1. Log in to Artemis
2. Go to your course detail page -> Statistics
3. check your current statistics
4. create an exercise, with start/due/assessment-date and participate in it, with different students
5. assess the submissions and keep in mind the score you give
6. Go back to course statistics:
6.1 check that corresponding graphs increased by the expected value (submissions, active users(only non-test user displayed), exercise releases, duedates, etc.)
6.2 Check that the average score of the exercise is correct (you can check the single scores under course management overview -> Exercises -> scores)
6.3 Check that you can click the right arrow if you have more than 10 exercises
6.4 Check that when hovering over the aver course score line (blue line in average score graph) that it matches the average score in _Participant Scores_ (Course detail page -> Participant Score -> on top it says average score: XX%)
7. Create questions and answers in a lecture and in an exercises in the course. The amount should be displayed in the _Questions Answered_ and _Questions Asked graph_

### Test Coverage



StatisticsResource.java: 100%
StatisticsService.java: 93%
StatisticsRepository.java: 85%
statistics-average-score-graph.component.ts: 80%
course-management-statistics.component.ts: 100%

### Screenshots


![Screenshot 2021-03-30 at 15 42 55](https://user-images.githubusercontent.com/44401560/112998763-a3795500-916e-11eb-9b1c-93921a0fe224.png)


Two new metrics:
<img width="1371" alt="Screenshot 2021-03-29 at 12 50 01" src="https://user-images.githubusercontent.com/44401560/112826491-54a7be80-908d-11eb-8764-ee8802a60448.png">
